### PR TITLE
WIP!

### PR DESCRIPTION
**Pull request by gilco-vastdata**
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 1]
```diff
@@ -3602,10 +3603,27 @@ private void createMergeAnalysis(Table table, TableHandle handle, TableSchema ta
                     .map(allColumnHandles::get)
                     .collect(toImmutableSet());
 
+            RowChangeParadigm paradigm = metadata.getRowChangeParadigm(session, handle);
+
             // create the RowType that holds all column values
             List<RowType.Field> fields = new ArrayList<>();
+            Set<ColumnHandle> updatedColumnsNamesSet =
```
Comment (rz-vastdata) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3610]:
```suggestion
            Set<ColumnHandle> updatedColumns =
```

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 2]
```diff
@@ -934,10 +942,17 @@ private MergeWriterNode createMergePipeline(Table table, RelationPlan relationPl
                     columnNamesBuilder.add(columnSchema.getName());
                 });
         MergeParadigmAndTypes mergeParadigmAndTypes = new MergeParadigmAndTypes(Optional.of(paradigm), typesBuilder.build(), columnNamesBuilder.build(), rowIdType);
-        MergeTarget mergeTarget = new MergeTarget(handle, Optional.empty(), metadata.getTableName(session, handle).getSchemaTableName(), mergeParadigmAndTypes);
+        MergeTarget mergeTarget = new MergeTarget(handle, Optional.empty(), metadata.getTableName(session, handle).getSchemaTableName(), mergeParadigmAndTypes, updatedColumnHandles);
 
         ImmutableList.Builder<Symbol> columnSymbolsBuilder = ImmutableList.builder();
-        for (ColumnHandle columnHandle : mergeAnalysis.getDataColumnHandles()) {
+        List<ColumnHandle> iterateHandles;
```
Comment (rz-vastdata) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:948]: `updatedColumnHandles.isEmpty()` can be true also when handling a DELETE, no?
```suggestion
        List<ColumnHandle> columnsHandles = (paradigm != RowChangeParadigm.UPDATE_PARTIAL_COLUMNS) ? mergeAnalysis.getDataColumnHandles() : updatedColumnHandles;
```

File: core/trino-main/src/main/java/io/trino/tracing/TracingConnectorMetadata.java [hunk 3]
```diff
@@ -733,11 +733,11 @@ public Optional<ConnectorPartitioningHandle> getUpdateLayout(ConnectorSession se
     }
 
     @Override
-    public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode)
+    public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode, List<ColumnHandle> updatedColumns)
     {
         Span span = startSpan("beginMerge", tableHandle);
         try (var ignored = scopedSpan(span)) {
-            return delegate.beginMerge(session, tableHandle, retryMode);
+            return delegate.beginMerge(session, tableHandle, retryMode, List.of());
```
Comment (rz-vastdata) [inline core/trino-main/src/main/java/io/trino/tracing/TracingConnectorMetadata.java:740]:
```suggestion
            return delegate.beginMerge(session, tableHandle, retryMode, updatedColumns);
```

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 4]
```diff
@@ -549,7 +551,9 @@ public PlanNode plan(Delete node)
                 assignmentsBuilder.putIdentity(symbol);
             }
             else {
-                assignmentsBuilder.put(symbol, new NullLiteral());
```
Comment (rz-vastdata) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:552]: Should we change this behaviour? If so, it is an optimization?

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java [hunk 5]
```diff
@@ -549,7 +551,9 @@ public PlanNode plan(Delete node)
                 assignmentsBuilder.putIdentity(symbol);
```
Comment (rz-vastdata) [inline core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:551]: Should we handle the redistribution case as well?

File: core/trino-main/src/test/java/io/trino/connector/MockConnector.java [hunk 6]
```diff
@@ -792,7 +792,7 @@ public ColumnHandle getMergeRowIdColumnHandle(ConnectorSession session, Connecto
         }
 
         @Override
-        public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode)
+        public ConnectorMergeTableHandle beginMerge(ConnectorSession session, ConnectorTableHandle tableHandle, RetryMode retryMode, List<ColumnHandle> updatedColumns)
```
Comment (rz-vastdata) [inline core/trino-main/src/test/java/io/trino/connector/MockConnector.java:795]: Consider adding a sanity test for the new merge type.

File: core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java (hunk 5 above)
Comment (gilco-vastdata) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:551]: Will add TODO

Comment (gilco-vastdata) [reply core/trino-main/src/main/java/io/trino/sql/planner/QueryPlanner.java:551]: `Kudu` and `Phoneix` does not utilize redistribution as well

**Comment by mosabua** [comment]
👋 @gilco-vastdata -- I am closing this PR as out of date. It is also missing any description or info that would allow us to review.

**Pull request closed without merging**